src/server/auth.ts

```diff
@@ -1,121 +1,183 @@
 
 /**
  * Module augmentation for `next-auth` types. Allows us to add custom properties to the `session`
  * object and keep type safety.
  *
  * @see https://next-auth.js.org/getting-started/typescript#module-augmentation
  */
+
+export type RoleName = "MENTOR" | "MENTEE" | "ADMIN";
+
 declare module "next-auth" {
   interface Session extends DefaultSession {
     user: DefaultSession["user"] & {
       id: string;
-      // ...other properties
-      // role: UserRole;
+      roleName: RoleName;
     };
   }
 
-  // interface User {
-  //   // ...other properties
-  //   // role: UserRole;
-  // }
+  interface User {
+    roleName: RoleName;
+  }
 }
 
-const prismaAdapter = PrismaAdapter(db);
-
 const credentialsAuthAvailable = () => process.env.NODE_ENV !== "production";
 
 /**
- * Options for NextAuth.js used to configure adapters, providers, callbacks, etc.
- *
- * @see https://next-auth.js.org/configuration/options
+ * In order to get credential authentication to work with NextAuth sessions,
+ * we need to wrap the NextAuth handler with this so we can manually
+ * create the session and store in db for credentials logins.
  */
-export const authOptions: NextAuthOptions = {
-  session: {
-    strategy: "jwt",
-  },
-  callbacks: {},
-  adapter: {
-    ...prismaAdapter,
-    async createUser(user) {
-      if (!user.name || !user.email) {
-        throw new Error("Missing name");
-      }
-      return db.user.create({
-        data: {
-          ...user,
-          name: user.name,
-          email: user.email,
-          emailVerified: new Date(Date.now()),
-          image: Picsum.url({ height: 128, cache: false }),
+export function requestWrapper(
+  req: NextApiRequest,
+  res: NextApiResponse,
+): [req: NextApiRequest, res: NextApiResponse, opts: NextAuthOptions] {
+  const generateSessionToken = () => randomUUID();
+
+  const fromDate = (time: number, date = Date.now()) =>
+    new Date(date + time * 1000);
+
+  const adapter = PrismaAdapter(db);
+
+  const opts: NextAuthOptions = {
+    cookies: {
+      sessionToken: {
+        name: "next-auth.session-token",
+        options: {
+          httpOnly: true,
+          secure: env.NODE_ENV === "production",
+          sameSite: "lax",
+          path: "/",
         },
-      });
+      },
     },
-  },
-  providers: [
-    GoogleProvider({
-      clientId: env.GOOGLE_CLIENT_ID,
-      clientSecret: env.GOOGLE_CLIENT_SECRET,
-    }),
-    ...(credentialsAuthAvailable()
-      ? [
-          CredentialsProvider({
-            name: "Credentials",
-            credentials: {
-              email: {
-                label: "Email",
-                type: "text",
-              },
-            },
-            async authorize(credentials) {
-              if (!credentials) {
-                return null;
-              }
-              const user = await db.user.findFirst({
-                where: {
-                  email: credentials.email,
+    adapter,
```

Comment:
This is stolen from https://github.com/nextauthjs/next-auth/discussions/4394#discussioncomment-3767702